refactor: Drop boost::thread and boost::chrono

ken2812221 opened the pull request:
Based on #14464 and #14480. Add interruptible thread class to replace boost::thread

The remaining usage of libboost_thread is boost::shared_mutex

ken2812221 committed:
refactor: Drop boost::this_thread::interruption_point and boost::thread_interrupted in main thread

ken2812221 committed:
utils: Add interruptible thread class

ken2812221 committed:
refactor: Replace boost::thread with InterruptibleThread

ken2812221 committed:
tests: Add interruptible thread tests

ken2812221 commented:
Apparently I wrote a MSVC-only code. I am going to investigate it.

MeshCollider commented:
Related #8631

ken2812221 committed:
msvc: Use wildcard on src files for bench_bitcoin.exe

ken2812221 committed:
build: Remove boost::chrono from build system

ken2812221 commented:
This is ready for review

practicalswift commented on src/threadinterrupt.cpp:
```diff
@@ -43,3 +43,30 @@ bool CThreadInterrupt::sleep_for(std::chrono::minutes rel_time)
+    if (g_interrupt_flag->m_interrupted) {
+        throw ThreadInterrupted();
+    }
+}
```
Missing newline at end of file :-)

ken2812221 commented:
@donaloconnor It should use mutex because it's used by the condition_variable.

For reference, you can find boost implementation for [pthread](https://github.com/boostorg/thread/blob/develop/src/pthread/thread.cpp), it acquires lock before it reads `interrupt_requested`.

donaloconnor commented:
@ken2812221 ah yes for the sleep. Thanks for the link to boosts impl.

ken2812221 commented:
Actually, I think it's fine to make it use both atomic and mutex at the same time. Then we don't have to lock mutex in InterruptionPoint.

donaloconnor commented:
In this case it might work but in practice it can cause race conditions. The value of the atomic can change from the moment the cond var is signalled and the predicate is checked (to work with spurious wakeups). Eg: false->true->notify+wakeup->false->predcheck.

The last false flag can never happen here (at least my understanding of the code) I guess so maybe it's okay.

ken2812221 commented:
@donaloconnor I am not sure people would allow this solution. Now the flag contains two variable:
- atomic bool read in `InterruptionPoint`
- non-atomic bool read in `InterruptionSleep`

donaloconnor commented:
I think it's not worth the complexity tbh. Sorry. I don't know if the perf improvement is worth the extra complexity without measurements. Maybe we should just stick to what boost did as you had in original. I think using the atomic in the cond var is okay for this case because it's state only goes from false->true.
I'd prefer if others chimed in to give feedback on this..
Re: spurious wakeups in the sleep. Do you think that's an issue? The sleep could get ended prematurely. Probably not a big deal. I don't know how much the code relies on accurate sleeps.

practicalswift commented:
What about splitting the work in two parts?

Start with the low-risk Boost replacement: that change is likely worth it in itself as long as no deviations from what Boost was doing are introduced.

When that work is finished: continue with the risker performance improvement change you are considering.

Does that make sense?

ken2812221 committed:
refactor: Make CCheckQueue interruptible

This commit add a Interrupt function for CCheckQueue that it can handle interrupt by itself instead of relying on boost thread interrupt

ken2812221 committed:
refactor: Drop boost in CCheckQueue

replace boost::mutex with debuggable Mutex
replace boost::condition_variable with std::condition_variable
add const specifier to fMaster and nBatchSize
add clang thread safety attributes
move init value of member of CCheckQueue from constructor to definition

ken2812221 committed:
refactor: Make CCheckQueue worker threads be owned by itself

Manage CCheckQueue workder threads by itself. Do not expose the threads outside the object.

MarcoFalke commented:
fyi, Boost::Chrono has been removed
